src/emulator/auth/operations.ts

```diff
@@ -2138,6 +2337,115 @@ function handleIdpSignUp(
   };
 }
 
+type MfaEnrollment = Schemas["GoogleCloudIdentitytoolkitV1MfaEnrollment"];
+
+interface MfaPendingCredential {
+  _AuthEmulatorMfaPendingCredential: string;
+  localId: string;
+  signInProvider: string;
+  projectId: string;
+}
+
+function mfaPending(
+  state: ProjectState,
+  user: UserInfo,
+  signInProvider: string
+): { mfaPendingCredential: string; mfaInfo: MfaEnrollment[] } {
+  if (!user.mfaInfo) {
+    throw new Error("Internal assertion error: mfaPending called on user without MFA.");
+  }
+  const pendingCredentialPayload: MfaPendingCredential = {
+    _AuthEmulatorMfaPendingCredential: "DO NOT MODIFY",
+    localId: user.localId,
+    signInProvider,
+    projectId: state.projectId, // TODO: Handle tenants.
+  };
+
+  // Encode pendingCredentialPayload using base64. We don't encrypt or sign the
+  // data in the Auth Emulator but just trust developers not to modify it.
+  const mfaPendingCredential = Buffer.from(
+    JSON.stringify(pendingCredentialPayload),
+    "utf8"
+  ).toString("base64");
+
+  return { mfaPendingCredential, mfaInfo: user.mfaInfo.map(redactMfaInfo) };
+}
+
+function redactMfaInfo(mfaInfo: MfaEnrollment): MfaEnrollment {
+  return {
+    displayName: mfaInfo.displayName,
+    enrolledAt: mfaInfo.enrolledAt,
+    mfaEnrollmentId: mfaInfo.mfaEnrollmentId,
+    phoneInfo: mfaInfo.unobfuscatedPhoneInfo
+      ? obfuscatePhoneNumber(mfaInfo.unobfuscatedPhoneInfo)
+      : undefined,
+  };
+}
+
+// Create an obfuscated version of a phone number, where all but the last
+// four digits are replaced by the character "*".
+function obfuscatePhoneNumber(phoneNumber: string): string {
+  let digitsLeft = 0;
```

Comment:
I think this function could be simpler. For example this should also accomplish the same thing:
```ts
function obfuscatePhoneNumber(phoneNumber: string) {
  const split = phoneNumber.split('');
  let updated = 0;
  for (let i = split.length - 1; i >= 0; i--) {
    if (/[0-9]/.test(split[i])) {
      updated++;
      split[i] = '*';
    }
    if (updated === 4) {
      break;
    }
  }
  return split.join('');
}
```

Reply:
Great idea, but your impl does the opposite -- fixed in PR